shared/src/main/scala/fetch.scala

```diff
@@ -17,207 +17,497 @@
-      Fetch.join(fa, fb).map { case (a, b) => f(a, b) }
+          val newResult = (r: FetchStatus) => x.result(r) >> y.result(r)
 
-    override def tuple2[A, B](fa: Fetch[A], fb: Fetch[B]): Fetch[(A, B)] =
-      Fetch.join(fa, fb)
+          BlockedRequest(newRequest, newResult)
+      }
   }
 
-  object Fetch extends FetchInterpreters {
+  /* A map from datasources to blocked requests used to group requests to the same data source. */
+  case class RequestMap(m: Map[DataSource[Any, Any], BlockedRequest])
 
-    /**
-     * Lift a plain value to the Fetch monad.
-     */
-    def pure[A](a: A): Fetch[A] =
-      Free.pure(a)
+  def optionCombine[A : Semigroup](a: A, opt: Option[A]): A =
+    opt.map(a |+| _).getOrElse(a)
 
-    /**
-     * Lift an exception to the Fetch monad.
-     */
-    def error[A](e: Throwable): Fetch[A] =
-      Free.liftF(Thrown(e))
+  implicit val rqSemigroup: Semigroup[RequestMap] = new Semigroup[RequestMap] {
+    def combine(x: RequestMap, y: RequestMap): RequestMap =
+      RequestMap(
+        x.m.foldLeft(y.m) {
+          case (acc, (ds, blocked)) => acc.updated(ds, optionCombine(blocked, acc.get(ds)))
+        }
+      )
+  }
 
-    /**
-     * Given a value that has a related `DataSource` implementation, lift it
-     * to the `Fetch` monad. When executing the fetch the data source will be
-     * queried and the fetch will return its result.
-     */
-    def apply[I, A](i: I)(implicit DS: DataSource[I, A]): Fetch[A] =
-      Free.liftF(FetchOne[I, A](i, DS))
+  // `Fetch` result data type
+  sealed trait FetchResult[A]
+  case class Done[A](x: A) extends FetchResult[A]
+  case class Blocked[A](rs: RequestMap, cont: Fetch[A]) extends FetchResult[A]
 
-    /**
-     * Given multiple values with a related `DataSource` lift them to the `Fetch` monad.
-     */
-    def multiple[I, A](i: I, is: I*)(implicit DS: DataSource[I, A]): Fetch[List[A]] =
-      Free.liftF[FetchOp, List[A]](FetchMany(NonEmptyList(i, is.toList), DS))
+  sealed trait Fetch[A] {
+    def run: IO[FetchResult[A]]
+  }
+  case class Unfetch[A](run: IO[FetchResult[A]]) extends Fetch[A]
 
-    /**
-     * Transform a list of fetches into a fetch of a list. It implies concurrent execution of fetches.
-     */
-    def sequence[I, A](ids: List[Fetch[A]]): Fetch[List[A]] =
-      traverse(ids)(identity)
+  implicit val fetchM: Monad[Fetch] = new Monad[Fetch] {
+    def pure[A](a: A): Fetch[A] =
+      Unfetch(
+        IO.pure(Done(a))
+      )
+
+    override def map[A, B](fa: Fetch[A])(f: A => B): Fetch[B] =
+      Unfetch(for {
+        fetch <- fa.run
+        result = fetch match {
+          case Done(v) => Done(f(v))
+          case Blocked(br, cont) =>
+            Blocked(br, map(cont)(f))
+        }
+      } yield result.asInstanceOf[FetchResult[B]])
+
+    override def product[A, B](fa: Fetch[A], fb: Fetch[B]): Fetch[(A, B)] =
+      Unfetch(for {
+        a <- fa.run
+        b <- fb.run
+        result = (a, b) match {
```

Comment:
Can this be expressed with an `apply` product? If it's possible I think it makes more sense.
```scala
(fa.run, fb.run).tupled.map {
  case (Done(a), Done(b)) => Done((a, b))
  //...
```